File: nsp_proselint/linter.py
"""Core linter engine: detect + replace cliche patterns."""

import re
import random
from dataclasses import dataclass
from typing import Optional

from .loader import DictionaryLoader, Dictionary, PatternEntry


@dataclass
class LintHit:
    clause: str
    pattern_id: str
    category: str
    match_text: str
    match_start: int
    match_end: int
    severity: str


@dataclass
class Diff:
    original: str
    replacement: str
    pattern_id: str
    clause_index: int


# Split after Chinese/English clause-ending punctuation.
# Keeps the delimiter attached to the preceding clause.
_CLAUSE_SPLIT_RE = re.compile(r"(?<=[，。！？；、.!?;,])")


def _is_attributive_context(match_text: str, clause: str, match_start: int) -> bool:
    """Check if match is in attributive context (ends with 的).

    When a regex match ends with "的", the greedy tail has consumed into a
    modifier structure (e.g. "极其细微的弧度", "极其细微的、近乎残忍的").
    A full replacement would break sentence structure. Skip these hits.
    """
    return match_text.endswith("的")


def _split_clauses(text: str) -> list[str]:
    """Split text into clauses by punctuation boundaries."""
    parts = _CLAUSE_SPLIT_RE.split(text)
    # Filter trailing empty string from split after final punctuation
    if parts and parts[-1] == "":
        parts = parts[:-1]
    return parts if parts else [""]


class ProseLinter:

    def __init__(self, lang: str = "zh_CN", dictionary: Optional[Dictionary] = None):
        if dictionary:
            self._dict = dictionary
        else:
            self._dict = DictionaryLoader.load_builtin(lang)

    @property
    def language(self) -> str:
        return self._dict.language

    @property
    def entries(self) -> list[PatternEntry]:
        return self._dict.entries

    def lint(self, text: str) -> list[LintHit]:
        """Detect cliches without replacing. Returns list of hits."""
        hits = []
        for para in text.split("\n"):
            if not para.strip():
                continue
            for clause in _split_clauses(para):
                if not clause.strip():
                    continue
                for entry in self._dict.entries:
                    m = entry.compiled.search(clause)
                    if m:
                        if _is_attributive_context(m.group(0), clause, m.start()):
                            continue
                        hits.append(LintHit(
                            clause=clause,
                            pattern_id=entry.id,
                            category=entry.category,
                            match_text=m.group(0),
                            match_start=m.start(),
                            match_end=m.end(),
                            severity=entry.severity,
                        ))
                        break  # first match per clause

        return hits

    def replace(self, text: str) -> str:
        """Detect and replace cliches. Returns cleaned text."""
        result, _ = self.replace_with_diff(text)
        return result

    def replace_with_diff(self, text: str) -> tuple[str, list[Diff]]:
        """Detect and replace cliches, tracking all changes.

        Returns (replaced_text, list_of_diffs).
        """
        diffs = []
        result_paragraphs = []
        clause_index = 0

        for para in text.split("\n"):
            if not para.strip():
                result_paragraphs.append(para)
                continue

            clauses = _split_clauses(para)
            result_clauses = []

            for clause in clauses:
                if not clause.strip():
                    result_clauses.append(clause)
                    continue

                replaced = False
                for entry in self._dict.entries:
                    m = entry.compiled.search(clause)
                    if m:
                        if _is_attributive_context(m.group(0), clause, m.start()):
                            continue
                        if entry.severity == "replace" and entry.replacements:
                            repl = random.choice(entry.replacements)
                            new_clause = clause[:m.start()] + repl + clause[m.end():]
                            diffs.append(Diff(
                                original=m.group(0),
                                replacement=repl,
                                pattern_id=entry.id,
                                clause_index=clause_index,
                            ))
                            result_clauses.append(new_clause)
                            replaced = True
                        # First match per clause, whether replaced or not
                        break

                if not replaced:
                    result_clauses.append(clause)
                clause_index += 1

            result_paragraphs.append("".join(result_clauses))

        return "\n".join(result_paragraphs), diffs

File: webapp/js/linter.js
/**
 * Core linter engine (JS port of nsp_proselint/linter.py).
 *
 * Loads dictionary JSON, compiles regex strings to RegExp,
 * splits text into clauses, detects cliche patterns with
 * per-clause first-match logic.
 */

const Linter = (() => {
    // Compiled dictionaries: [{id, category, type, severity, detect, regex, replacements, suffix_type?}]
    let _dictionaries = [];

    // Split after Chinese/English clause-ending punctuation.
    // Keeps delimiter attached to preceding clause.
    const CLAUSE_SPLIT_RE = /(?<=[，。！？；、.!?;,])/;

    function splitClauses(text) {
        const parts = text.split(CLAUSE_SPLIT_RE);
        // Filter trailing empty string
        if (parts.length > 0 && parts[parts.length - 1] === "") {
            parts.pop();
        }
        return parts.length > 0 ? parts : [""];
    }

    /**
     * Load and compile a dictionary from JSON array.
     * @param {Array} entries - Raw entries from JSON
     * @returns {Array} Compiled entries with regex property
     */
    function compileDictionary(entries) {
        return entries.map(e => ({
            ...e,
            regex: new RegExp(e.detect),
        }));
    }

    /**
     * Load dictionaries from JSON URLs.
     * @param {string[]} urls - Array of JSON file URLs
     */
    async function loadDictionaries(urls) {
        _dictionaries = [];
        const results = await Promise.all(
            urls.map(url => fetch(url).then(r => r.json()))
        );
        for (const entries of results) {
            _dictionaries.push(...compileDictionary(entries));
        }
        return _dictionaries.length;
    }

    /**
     * Get loaded dictionaries.
     */
    function getDictionaries() {
        return _dictionaries;
    }

    /**
     * Lint a single paragraph, returning hits with paragraph-relative offsets.
     *
     * @param {string} para - One paragraph of text
     * @returns {Array} hits: [{entryId, category, type, severity, matchText,
     *                          start, end, replacements, suffixType, clauseIndex}]
     */
    function lintParagraph(para) {
        if (!para.trim()) return [];

        const hits = [];
        const clauses = splitClauses(para);
        let offset = 0; // cumulative offset for paragraph-relative positions

        for (let ci = 0; ci < clauses.length; ci++) {
            const clause = clauses[ci];
            if (!clause.trim()) {
                offset += clause.length;
                continue;
            }

            for (const entry of _dictionaries) {
                const m = entry.regex.exec(clause);
                if (m) {
                    // Skip if match ends with "的" followed by non-punctuation
                    // (attributive structure like "细微的弧度" — replacement would break syntax)
                    if (_isAttributiveContext(m[0])) {
                        continue;
                    }

                    hits.push({
                        entryId: entry.id,
                        category: entry.category,
                        type: entry.type,
                        severity: entry.severity,
                        matchText: m[0],
                        start: offset + m.index,
                        end: offset + m.index + m[0].length,
                        replacements: entry.replacements || [],
                        suffixType: entry.suffix_type || null,
                        clauseIndex: ci,
                    });
                    break; // first match per clause
                }
            }
            offset += clause.length;
        }

        return hits;
    }

    /**
     * Lint entire text (multiple paragraphs).
     * Returns array of {paraIndex, hits[]}, one per non-empty paragraph.
     */
    function lintText(paragraphs) {
        const results = [];
        for (let pi = 0; pi < paragraphs.length; pi++) {
            const hits = lintParagraph(paragraphs[pi]);
            if (hits.length > 0) {
                results.push({ paraIndex: pi, hits });
            }
        }
        return results;
    }

    /**
     * Check if a match is in attributive context: ends with "的".
     * When a greedy regex tail consumes into a modifier structure
     * (e.g. "极其细微的弧度", "极其细微的、"), a full replacement
     * would break sentence structure. Skip these hits.
     */
    function _isAttributiveContext(matchText) {
        return matchText.endsWith("的");
    }

    return {
        splitClauses,
        compileDictionary,
        loadDictionaries,
        getDictionaries,
        lintParagraph,
        lintText,
    };
})();
